### ISSUE TITLE
with_min_len/with_max_len doesn't work with u64 ranges because they don't implement the indexed trait

### ISSUE THREAD

**Issue by winsut** [nikomatsakis/rayon#359]
This is highly surprising and I think it definitely should not work like this.

My guess is that u64 should be used instead of usize everywhere, since programs can process more than 2^32 iterator items (but not more than 2^64 due to limited CPU speeds and human lifetimes).

**Comment by cuviper** [comment]
Note that the standard library has `ExactSizeIterator` with the method `fn len(&self) -> usize`, which is not implemented by `Range<u64>`.  I think it would be _more_ surprising if Rayon departed from the established pattern of using `usize` lengths.

That said, I agree it would be nice to improve the support for 64-bit ranges, somehow.

**Comment by winsut** [comment]
There might be a design mistake in the standard library too.

That said, the standard library mostly/only uses ExactSizeIterator to pre-allocate space for collecting from an iterator, and by definition you can't do that for values greater than usize::max_value().

On the hand, Rayon, unlike the standard library, uses it to split the iterator, and here u64 makes sense, since a 32-bit program can still execute more than 2^32 data items (but probably not more than 2^64, at least with current technology) even though it can't store them in memory.

There could be an argument that it should actually be u128, since if Rayon were to add support for distributing task over a network, it would become possible to run 2^64-2^128 tasks.

**Comment by cuviper** [comment]
You can accomplish *any* bit-depth by nesting iterators, like:

```rust
    let last = (0u32..1_000_000_000)
        .into_par_iter()
        .map(|i| i as u128 * 1_000_000_000)
        .flat_map(
            |i| {
                (0u32..1_000_000_000)
                    .into_par_iter()
                    .map(move |j| i + j as u128)
            }
        )
        .map(|i| i as u128 * 1_000_000_000)
        .flat_map(
            |i| {
                (0u32..1_000_000_000)
                    .into_par_iter()
                    .map(move |j| i + j as u128)
                    .with_min_len(1_000)
                    .with_max_len(1_000_000)
            }
        )
        .find_last(|_| true);
    println!("{:?}", last);
```

That's iterating 0..10^27, with each sequential job limited between a thousand and a million items, and this correctly reports `Some(999999999999999999999999999)`.  It doesn't even take very long since `find_last` prunes most of the lesser splits as soon as anything is found.

You can still argue that this is harder to set up than it should be, but it does work.

**Comment by nikomatsakis** [comment]
To what extent is our use of `usize` something that the "end-user sees" versus an implementation detail? 

One thing that maybe comes to mind is that, when we switch to the "sequential iteration" phase, we currently do have some annoyances around `ExactSizeIterator`, although I think that's kind of orthogonal from this (that's more to do w/ `rev()` and `zip()`, I guess). But not wholly independent.

Another question, could we just make `u64` ranges work but panic dynamically if they span a gap bigger than `usize::MAX` and use `with_min_len` or whatever?

**Comment by cuviper** [comment]
> To what extent is our use of usize something that the "end-user sees" versus an implementation detail?

`len()`, `enumerate()`, `position_any()` et al., and indirectly `collect_into()`, all expose `usize`.

> Another question, could we just make u64 ranges work but panic dynamically if they span a gap bigger than usize::MAX and use with_min_len or whatever?

We could assert in `len()` -- we already cheat that way with `chain()`.  But that doesn't satisfy @winsut's desire to actually iterate more than `usize::MAX` items.  Or if you want to actually use `u64` lengths internally, then we'd have to assert on all the other `usize`-ish APIs.

**Comment by nikomatsakis** [comment]
Yes, OK. An annoying problem. I do certainly agree that e.g. `enumerate()` should match the `Iterator` signature. It *may* be worthwhile trying to convert the "internal" APIs to use `u64`, though. Unclear.

**Comment by nikomatsakis** [comment]
Discussing in the meeting. I think the inclination is not to solve this for 1.0 -- even if we found a satisfactory solution, it wouldn't have much time to "Bake" and it involves tinkering at a fairly low level with the `IndexedParallelIterator` trait and so forth.

We can mitigate by adding inherent methods to `Range<u64>` and would consider fixing this either via a 2.0 release someday or if default associated types or some other language feature are added to rust that makes it more plausible.
